# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [3]:
women = train_df.loc[train_df.Sex == 'female'].loc[train_df.Pclass == 1]["Survived"]
rate_women = sum(women)/len(women)

print("% of women who survived:", rate_women)



% of women who survived: 0.9680851063829787


# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [4]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [5]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [6]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2025-01-27T21:36:55.484214518+00:00 kernel.cc:1214] Loading model from path /tmp/tmplvtrdiy4/model/ with prefix 97ca7fc8ca8c4952
[INFO 2025-01-27T21:36:55.495009672+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:36:55.495227448+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [8]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2025-01-27T21:36:58.921876393+00:00 kernel.cc:1214] Loading model from path /tmp/tmpghdmxafb/model/ with prefix e1ccda34bfae4ed1
[INFO 2025-01-27T21:36:58.932539893+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2025-01-27T21:36:58.932599689+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [9]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [10]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [11]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmphfxc9mzx as temporary training directory


[INFO 2025-01-27T21:39:03.416777333+00:00 kernel.cc:1214] Loading model from path /tmp/tmphfxc9mzx/model/ with prefix 6ee9ddc870d84a0b
[INFO 2025-01-27T21:39:03.432832185+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2025-01-27T21:39:03.433009416+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2025-01-27T21:39:03.433090235+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [12]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2025-01-27T21:39:04.422456089+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf30jvlof/model/ with prefix 97e320fde0be4ddf
[INFO 2025-01-27T21:39:04.427474983+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2025-01-27T21:39:05.815978859+00:00 kernel.cc:1214] Loading model from path /tmp/tmprvr1jmc5/model/ with prefix 9aea420847594472
[INFO 2025-01-27T21:39:05.83789457+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2025-01-27T21:39:06.822627944+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_iktnscl/model/ with prefix bb574da473564010
[INFO 2025-01-27T21:39:06.827811932+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2025-01-27T21:39:08.39596716+00:00 kernel.cc:1214] Loading model from path /tmp/tmpstetftdj/model/ with prefix 6f675379d117495e
[INFO 2025-01-27T21:39:08.425206045+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2025-01-27T21:39:09.420307916+00:00 kernel.cc:1214] Loading model from path /tmp/tmprc01jkil/model/ with prefix 7513a5f97ff64309
[INFO 2025-01-27T21:39:09.426670117+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2025-01-27T21:39:10.352826418+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4fauh8xq/model/ with prefix e5860f4d1b5c4c81
[INFO 2025-01-27T21:39:10.356341255+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2025-01-27T21:39:11.422259577+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzrr_z4ls/model/ with prefix 532e36d13f414da7
[INFO 2025-01-27T21:39:11.430864377+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2025-01-27T21:39:12.835108574+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0rc0shea/model/ with prefix a3fc7c61e8c24c08
[INFO 2025-01-27T21:39:12.857826809+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2025-01-27T21:39:14.076477171+00:00 kernel.cc:1214] Loading model from path /tmp/tmpznv0oegq/model/ with prefix 89ddf110d6e247da
[INFO 2025-01-27T21:39:14.087774567+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:39:14.087836443+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2025-01-27T21:39:15.77915432+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_n6y3lbd/model/ with prefix 3b39095a149648ae
[INFO 2025-01-27T21:39:15.797120455+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2025-01-27T21:39:16.831433086+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_a4qqae4/model/ with prefix fbfbe0f0635b42c7
[INFO 2025-01-27T21:39:16.837917126+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2025-01-27T21:39:18.078327359+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmx66jos2/model/ with prefix 16d5cda173744876
[INFO 2025-01-27T21:39:18.094553701+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2025-01-27T21:39:19.096189523+00:00 kernel.cc:1214] Loading model from path /tmp/tmpeousbb8b/model/ with prefix 99e6ca89ba86486d
[INFO 2025-01-27T21:39:19.103250043+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2025-01-27T21:39:20.246631792+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjx5kam75/model/ with prefix 5904a9540afe4f3d
[INFO 2025-01-27T21:39:20.259314986+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2025-01-27T21:39:21.298354149+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0x_1mz00/model/ with prefix e17161ee22c94784
[INFO 2025-01-27T21:39:21.305650059+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2025-01-27T21:39:22.374377439+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6o7mrv6g/model/ with prefix cbf6f6713d544c04
[INFO 2025-01-27T21:39:22.383291375+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2025-01-27T21:39:23.571077438+00:00 kernel.cc:1214] Loading model from path /tmp/tmphqpz7ddf/model/ with prefix 10b3fd883af14b73
[INFO 2025-01-27T21:39:23.585480525+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2025-01-27T21:39:24.793478835+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjpez0fs8/model/ with prefix 479e4597215045db
[INFO 2025-01-27T21:39:24.808419586+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:39:24.808484149+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2025-01-27T21:39:25.962360167+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_57shme3/model/ with prefix 8b2b78f416a44688
[INFO 2025-01-27T21:39:25.975404838+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2025-01-27T21:39:27.338862341+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsciclowb/model/ with prefix 053941f8c09643a5
[INFO 2025-01-27T21:39:27.359051848+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2025-01-27T21:39:28.625699233+00:00 kernel.cc:1214] Loading model from path /tmp/tmpua58goe0/model/ with prefix 7b5674e624d54499
[INFO 2025-01-27T21:39:28.642539458+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2025-01-27T21:39:29.618340251+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0mvdmtzm/model/ with prefix 9849caa8e51648e6
[INFO 2025-01-27T21:39:29.624377245+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2025-01-27T21:39:30.61673191+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl9vdfwre/model/ with prefix da4091053a3c4736
[INFO 2025-01-27T21:39:30.623198156+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2025-01-27T21:39:31.775199184+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxezv9uge/model/ with prefix 4b72000f0b8443f5
[INFO 2025-01-27T21:39:31.785111814+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2025-01-27T21:39:32.808369194+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_vkjih8l/model/ with prefix fad4dbf504fb4c1e
[INFO 2025-01-27T21:39:32.815940477+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2025-01-27T21:39:34.008716989+00:00 kernel.cc:1214] Loading model from path /tmp/tmplflb4s95/model/ with prefix 6a06193199214927
[INFO 2025-01-27T21:39:34.021697088+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2025-01-27T21:39:35.163854212+00:00 kernel.cc:1214] Loading model from path /tmp/tmptk8dj42z/model/ with prefix 9951ee5a623c4938
[INFO 2025-01-27T21:39:35.175103463+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:39:35.175168734+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2025-01-27T21:39:36.19938861+00:00 kernel.cc:1214] Loading model from path /tmp/tmplxkg8yg9/model/ with prefix 65948de5cdbd4821
[INFO 2025-01-27T21:39:36.207077459+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2025-01-27T21:39:37.192698664+00:00 kernel.cc:1214] Loading model from path /tmp/tmp__g1r1sw/model/ with prefix 88d4a334b49d4c99
[INFO 2025-01-27T21:39:37.198715754+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2025-01-27T21:39:38.40277888+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyhkzyepz/model/ with prefix 034ba3b0695e46e2
[INFO 2025-01-27T21:39:38.417493423+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2025-01-27T21:39:40.017804282+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4rme6a4e/model/ with prefix eeb0dab1bdf04f68
[INFO 2025-01-27T21:39:40.048133761+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2025-01-27T21:39:41.751628605+00:00 kernel.cc:1214] Loading model from path /tmp/tmp304v_j9i/model/ with prefix 27d18120f29b4255
[INFO 2025-01-27T21:39:41.764756673+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2025-01-27T21:39:42.832087265+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6_v55q5m/model/ with prefix 67361103ee674c22
[INFO 2025-01-27T21:39:42.839567482+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2025-01-27T21:39:44.126990408+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy7otg5o6/model/ with prefix 55f378ef5b464ee8
[INFO 2025-01-27T21:39:44.142538246+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2025-01-27T21:39:45.309587173+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcoms5dp9/model/ with prefix d27bac178aaf4c19
[INFO 2025-01-27T21:39:45.323669796+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:39:45.323721006+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2025-01-27T21:39:46.464296928+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9z9qwlha/model/ with prefix 88277cb069614dcd
[INFO 2025-01-27T21:39:46.473708284+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2025-01-27T21:39:47.755699049+00:00 kernel.cc:1214] Loading model from path /tmp/tmpinvli0hl/model/ with prefix 025e5c7e83d74f21
[INFO 2025-01-27T21:39:47.771993505+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2025-01-27T21:39:48.894979582+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2m0xsiy2/model/ with prefix e83e41b25afc49a5
[INFO 2025-01-27T21:39:48.90522389+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2025-01-27T21:39:50.144290927+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdxfftng_/model/ with prefix 586610dd6b0147e5
[INFO 2025-01-27T21:39:50.159314882+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2025-01-27T21:39:51.415354894+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq5l8fywm/model/ with prefix 1860823b3f264268
[INFO 2025-01-27T21:39:51.429971151+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2025-01-27T21:39:52.425329771+00:00 kernel.cc:1214] Loading model from path /tmp/tmppaz26660/model/ with prefix 3eb06a0011c548e0
[INFO 2025-01-27T21:39:52.430804954+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2025-01-27T21:39:53.729815003+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdi54tjyt/model/ with prefix a5fb3eb69a814201
[INFO 2025-01-27T21:39:53.747159756+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2025-01-27T21:39:54.880922766+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb_d9dxwm/model/ with prefix b20f4c5f9eb0471b
[INFO 2025-01-27T21:39:54.891676908+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2025-01-27T21:39:56.248400579+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbjmjaazs/model/ with prefix 9d4acfa6340b4c8e
[INFO 2025-01-27T21:39:56.267822124+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:39:56.267873428+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2025-01-27T21:39:57.39163582+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2976g1oq/model/ with prefix 88206300315b45ec
[INFO 2025-01-27T21:39:57.402429415+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2025-01-27T21:39:58.340676547+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd4aq5psj/model/ with prefix 59237629ae1845d8
[INFO 2025-01-27T21:39:58.344917293+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2025-01-27T21:39:59.612322958+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1wz8xcmz/model/ with prefix 2bb45dc19db14f38
[INFO 2025-01-27T21:39:59.628587865+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2025-01-27T21:40:00.912226892+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxiuu0xou/model/ with prefix 85fbc9817143459d
[INFO 2025-01-27T21:40:00.927121418+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2025-01-27T21:40:01.931866192+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyj6ktffn/model/ with prefix 11f77e8857364714
[INFO 2025-01-27T21:40:01.937714564+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2025-01-27T21:40:02.99027917+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx2dcw9t4/model/ with prefix 2a34aaf6445b48c6
[INFO 2025-01-27T21:40:02.997869736+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2025-01-27T21:40:04.187816863+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgzix2viq/model/ with prefix 21a7387fd9d4447b
[INFO 2025-01-27T21:40:04.202072476+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2025-01-27T21:40:05.494486564+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0ppjiuef/model/ with prefix 729393943adb4380
[INFO 2025-01-27T21:40:05.511900125+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2025-01-27T21:40:06.619217026+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp_n3s7wo/model/ with prefix 6207ff49ee294fb2
[INFO 2025-01-27T21:40:06.627924445+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:40:06.627970325+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2025-01-27T21:40:07.72387755+00:00 kernel.cc:1214] Loading model from path /tmp/tmpohxanges/model/ with prefix 2e5d7ec72cbc4617
[INFO 2025-01-27T21:40:07.732463632+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2025-01-27T21:40:08.695041977+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdq7kqyov/model/ with prefix 14cd5c7cd5154d28
[INFO 2025-01-27T21:40:08.698913531+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2025-01-27T21:40:09.990482661+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4zymsq_6/model/ with prefix 2d3c0919123f4d21
[INFO 2025-01-27T21:40:10.006661089+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2025-01-27T21:40:11.82698401+00:00 kernel.cc:1214] Loading model from path /tmp/tmpi83aaseq/model/ with prefix 4c898beaee174e57
[INFO 2025-01-27T21:40:11.841797682+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2025-01-27T21:40:12.90286357+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7oqdoq1r/model/ with prefix 703c4e1f9d55451d
[INFO 2025-01-27T21:40:12.908049485+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2025-01-27T21:40:13.987133064+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqazfhrzs/model/ with prefix 0209dc6172d44d56
[INFO 2025-01-27T21:40:13.994746745+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2025-01-27T21:40:15.160485131+00:00 kernel.cc:1214] Loading model from path /tmp/tmp64exsmad/model/ with prefix dd08eb2e9f3648ea
[INFO 2025-01-27T21:40:15.171243539+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2025-01-27T21:40:16.308751319+00:00 kernel.cc:1214] Loading model from path /tmp/tmpljtq96vo/model/ with prefix 62acbd1257ba4789
[INFO 2025-01-27T21:40:16.320063103+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2025-01-27T21:40:17.412665691+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_gp_1pok/model/ with prefix 6fb7e5b2103c400b
[INFO 2025-01-27T21:40:17.418622407+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:40:17.418681555+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2025-01-27T21:40:19.048239343+00:00 kernel.cc:1214] Loading model from path /tmp/tmpivkam9_q/model/ with prefix 0b19b18ed354442b
[INFO 2025-01-27T21:40:19.077570294+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2025-01-27T21:40:20.204158572+00:00 kernel.cc:1214] Loading model from path /tmp/tmph7dmdvey/model/ with prefix 7f16b152c80245ca
[INFO 2025-01-27T21:40:20.214465551+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2025-01-27T21:40:21.35537222+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2v8e7nzh/model/ with prefix b7a6462461d642ec
[INFO 2025-01-27T21:40:21.364830199+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2025-01-27T21:40:22.383163707+00:00 kernel.cc:1214] Loading model from path /tmp/tmpa210ouyu/model/ with prefix 5ce7024af91a45e2
[INFO 2025-01-27T21:40:22.388714518+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2025-01-27T21:40:23.420179266+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbjzh4jyt/model/ with prefix 510a01c3f3f04d21
[INFO 2025-01-27T21:40:23.427635477+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2025-01-27T21:40:24.839379445+00:00 kernel.cc:1214] Loading model from path /tmp/tmpz03pjlyk/model/ with prefix e8ee3d2e76514c90
[INFO 2025-01-27T21:40:24.860827347+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2025-01-27T21:40:26.061455142+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw8w69a_s/model/ with prefix 9528337ee9734445
[INFO 2025-01-27T21:40:26.075078057+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2025-01-27T21:40:27.162690123+00:00 kernel.cc:1214] Loading model from path /tmp/tmpta99m29l/model/ with prefix d18cf88f8fea406d
[INFO 2025-01-27T21:40:27.170212004+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2025-01-27T21:40:28.279368697+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5kkpmxdt/model/ with prefix 34d98ba2209041f4
[INFO 2025-01-27T21:40:28.289235847+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:40:28.289286655+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2025-01-27T21:40:29.348476405+00:00 kernel.cc:1214] Loading model from path /tmp/tmptp81o32c/model/ with prefix 6b35ad679f434dfb
[INFO 2025-01-27T21:40:29.357147891+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2025-01-27T21:40:30.654177178+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1k558jgs/model/ with prefix 3ad7d946a9814e73
[INFO 2025-01-27T21:40:30.673273271+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2025-01-27T21:40:31.771922132+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxx3y1k5z/model/ with prefix ae04945b32b34bc1
[INFO 2025-01-27T21:40:31.7800747+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2025-01-27T21:40:32.991588955+00:00 kernel.cc:1214] Loading model from path /tmp/tmp41jdkfo9/model/ with prefix 29d9123a3c584942
[INFO 2025-01-27T21:40:33.004887825+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2025-01-27T21:40:34.106782491+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5jy8ropf/model/ with prefix 776fb9f7b44d4b8f
[INFO 2025-01-27T21:40:34.116123398+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2025-01-27T21:40:35.115019825+00:00 kernel.cc:1214] Loading model from path /tmp/tmpge00hwvj/model/ with prefix 176a3855d9904f1f
[INFO 2025-01-27T21:40:35.119761395+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2025-01-27T21:40:36.125004368+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmjvs0f6z/model/ with prefix 3b65c950c0684b59
[INFO 2025-01-27T21:40:36.130364021+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2025-01-27T21:40:37.169563031+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbfn_0n4e/model/ with prefix 50d80615c9254a08
[INFO 2025-01-27T21:40:37.177207382+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2025-01-27T21:40:38.212790244+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcdyg1_zt/model/ with prefix b8c234350ff44726
[INFO 2025-01-27T21:40:38.220787273+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2025-01-27T21:40:39.382597038+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2s9_wc27/model/ with prefix cd4ffdf2483e42c9
[INFO 2025-01-27T21:40:39.394604034+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:40:39.394656782+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2025-01-27T21:40:40.609727735+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsepo5x0y/model/ with prefix 5c5f31f48b72453e
[INFO 2025-01-27T21:40:40.622527653+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2025-01-27T21:40:41.80730199+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvdm9f69w/model/ with prefix 630606524f3b4c95
[INFO 2025-01-27T21:40:41.818743827+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2025-01-27T21:40:43.01056951+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2hesl_jw/model/ with prefix c7355698845d45ce
[INFO 2025-01-27T21:40:43.020860638+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2025-01-27T21:40:45.215404579+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo5ev62gn/model/ with prefix 36052efe3ec24b41
[INFO 2025-01-27T21:40:45.236814531+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2025-01-27T21:40:46.375079992+00:00 kernel.cc:1214] Loading model from path /tmp/tmpte4rksch/model/ with prefix 3fc3082b3b344018
[INFO 2025-01-27T21:40:46.383175731+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2025-01-27T21:40:47.83156191+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu59lyg0p/model/ with prefix 4f9512c25b334876
[INFO 2025-01-27T21:40:47.850553931+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2025-01-27T21:40:49.374330475+00:00 kernel.cc:1214] Loading model from path /tmp/tmp48b_3j1l/model/ with prefix 1c9345664f6e4de9
[INFO 2025-01-27T21:40:49.395854482+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:40:49.395923048+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2025-01-27T21:40:50.643469387+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsab1iw6h/model/ with prefix 418ba04d44dc46f5
[INFO 2025-01-27T21:40:50.656720142+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2025-01-27T21:40:51.74399723+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdm11llky/model/ with prefix 3efb79c833f64007
[INFO 2025-01-27T21:40:51.74968731+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2025-01-27T21:40:53.015010425+00:00 kernel.cc:1214] Loading model from path /tmp/tmp86gnupkp/model/ with prefix 40b84213755d4ccf
[INFO 2025-01-27T21:40:53.026877384+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2025-01-27T21:40:54.132308854+00:00 kernel.cc:1214] Loading model from path /tmp/tmp83k1os1c/model/ with prefix e7d2482e055a4d92
[INFO 2025-01-27T21:40:54.140220386+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2025-01-27T21:40:55.576206859+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2btvaozg/model/ with prefix 4f9f22e020484882
[INFO 2025-01-27T21:40:55.596800542+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2025-01-27T21:40:56.779678161+00:00 kernel.cc:1214] Loading model from path /tmp/tmplyhp9noj/model/ with prefix 045fd3b844e84bac
[INFO 2025-01-27T21:40:56.791338042+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2025-01-27T21:40:57.893897332+00:00 kernel.cc:1214] Loading model from path /tmp/tmpeg5l32k5/model/ with prefix cdf3d8415c854cc8
[INFO 2025-01-27T21:40:57.90157933+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2025-01-27T21:40:59.05954589+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfi0i6f9c/model/ with prefix 38a51081e9e74d8c
[INFO 2025-01-27T21:40:59.069071042+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2025-01-27T21:41:00.267387304+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbbwffpxl/model/ with prefix 6337687daa7a4eb0
[INFO 2025-01-27T21:41:00.278790608+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2025-01-27T21:41:00.278838861+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2025-01-27T21:41:01.359309253+00:00 kernel.cc:1214] Loading model from path /tmp/tmpt0sp8ee3/model/ with prefix 333cef0181654fe7
[INFO 2025-01-27T21:41:01.365326461+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2025-01-27T21:41:02.479408466+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvfttlfdm/model/ with prefix abff122c1c4447d8
[INFO 2025-01-27T21:41:02.488260853+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2025-01-27T21:41:03.801156634+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu0fd_ney/model/ with prefix 8ed75b5f792b4416
[INFO 2025-01-27T21:41:03.817820683+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 